numpy/core/arrayprint.py

```diff
@@ -521,6 +509,8 @@ def array2string(a, max_line_width=None, precision=None,
     if formatter is None:
         formatter = _formatter
 
+    if style is not None and a.shape == () and a.dtype.names is None:
+        return style(a[()])
```

Comment:
For discussion, that would be counter to current numpy behavior for all scalar types. That is, currently all 0d arrays are affected by `np.setprintfunction` including string types.

I agree that's somewhat counter to the idea of this PR. In this PR, we are making the `str` of 0d arrays behave more like scalars, and the `repr` more like ndarrays. You are suggesting we make the `str` even more like scalars, by not being affected by `np.setprintfunction`.

In some sense I agree, but on the other hand I am hesitant to change even more behavior, and I think the 0d behavior (even in this PR) is a little strange anyway.

I'm going to be away next week and was hoping to get this PR in before then. Do you mind leaving this point for consideration in another PR?